pkg/vm/engine/tae/logtail/service/server.go

```diff
@@ -0,0 +1,465 @@
+		s.options.logger = logger
+	}
+}
+
+// WithServerMaxMessageSize sets max rpc message size
+func WithServerMaxMessageSize(maxMessageSize int) ServerOption {
+	return func(s *LogtailServer) {
+		s.options.maxMessageSize = maxMessageSize
+	}
+}
+
+// WithServerEnableChecksum enables checksum
+func WithServerEnableChecksum(enable bool) ServerOption {
+	return func(s *LogtailServer) {
+		s.options.enableChecksum = enable
+	}
+}
+
+// WithCollectInterval sets logtail collection interval.
+func WithCollectInterval(interval time.Duration) ServerOption {
+	return func(s *LogtailServer) {
+		s.options.collectInterval = interval
+	}
+}
+
+// tableLogtail describes a table's logtail.
+type tableLogtail struct {
+	id   TableID
+	tail *logtail.TableLogtail
+}
+
+// publishment describes a batch of logtail.
+type publishment struct {
+	want  timestamp.Timestamp
+	tails []tableLogtail
+}
+
+// sessionError describes error when writing via morpc stream.
+type sessionError struct {
+	session *Session
+	err     error
+}
+
+// subscription describes new subscription.
+type subscription struct {
+	sendCtx context.Context
+	tableID TableID
+	req     *logtail.SubscribeRequest
+	session *Session
+}
+
+// LogtailServer handles logtail push logic.
+type LogtailServer struct {
+	pool struct {
+		requests  *sync.Pool
+		responses *sync.Pool
+	}
+
+	options struct {
+		logger          *zap.Logger
+		maxMessageSize  int
+		enableChecksum  bool
+		collectInterval time.Duration
+	}
+
+	ssmgr      *SessionManager
+	subscribed *Waterliner
+
+	errChan chan sessionError // errChan has no buffer in order to improve sensitivity.
+	pubChan chan publishment
+	subChan chan subscription
+
+	fetcher   LogtailFetcher
+	timestamp Timestamp
+
+	rpc morpc.RPCServer
+
+	rootCtx    context.Context
+	cancelFunc context.CancelFunc
+}
+
+// NewLogtailServer initializes a api server.
+func NewLogtailServer(
+	address string, fetcher LogtailFetcher, timestamp Timestamp, opts ...ServerOption,
+) (morpc.RPCServer, error) {
+	s := &LogtailServer{
+		ssmgr:      NewSessionManager(),
+		subscribed: NewWaterliner(),
+		errChan:    make(chan sessionError),
+		pubChan:    make(chan publishment),
+		subChan:    make(chan subscription, 10),
+		fetcher:    fetcher,
+		timestamp:  timestamp,
+	}
+
+	s.pool.requests = &sync.Pool{
+		New: func() any {
+			return &LogtailRequest{}
+		},
+	}
+	s.pool.responses = &sync.Pool{
+		New: func() any {
+			return &LogtailResponse{}
+		},
+	}
+
+	// default configuration
+	s.options.maxMessageSize = 16 * 1024
+	s.options.logger = logutil.GetGlobalLogger().
+		Named(LogtailServiceRPCName).
+		With(zap.String("uuid", uuid.NewString()))
+	s.options.enableChecksum = true
+	s.options.collectInterval = 50 * time.Millisecond
+
+	for _, opt := range opts {
+		opt(s)
+	}
+
+	codec := morpc.NewMessageCodec(s.acquireRequest,
+		morpc.WithCodecPayloadCopyBufferSize(s.options.maxMessageSize),
+		morpc.WithCodecEnableChecksum(),
+		morpc.WithCodecMaxBodySize(1024*1024),
```

Comment:
s.options.maxMessageSize set to payload copy buffer size?

Reply:
typo, resolved in the following commit.